### DISCUSSION TITLE
Attach additional metadata to dynamic partitions

### DISCUSSION THREAD

**Discussion by f-tremblay** [dagster-io/dagster#13878] [Ideas]
When creating runs for dynamic partitions, there is currently no way to define additional metadata other than the partition_key.

For some use cases, a single value is not enough to define a partition. 
An example is dynamic time ranges:  each partition is defined by an arbitrary _start_datetime_ and _end_datetime_ (and booleans _include_start_ and _include_end_). Another example is car _make_ and _model_. Another example is ML experiment parameters.

In these situations, the user is forced to concatenate them (e.g. `partition_key = f"{metadata_value_1}--{metadata_value_2}--{metadata_value_3}"`), and parse them inside the asset to extract the values. This workaround is not very convenient (you have to handle the parsing), or robust (there's no validation for the types or for missing values, etc.).
 
Another workaround would be to define a Pydantic model to represent the partition metadata, and use (de)-serialization like `my_partition_key = str(my_model.json())` and `my_model = MyModel.parse_raw(my_partition_key)`. This workaround is more convenient and robust. However, it produces long partition names, which might not be ideal in the UI.

Ideally, Dagster would provide support for attaching additional metadata to a dynamic partition (or to the RunRequest).
This way, the partition_key would not have to be overloaded with a very long value (like with workaround # 2), and the additionnal metadata values could be displayed and plotted in Dagit partitions tab.
RunRequest would take an additional argument (e.g. a dict, a dataclass or a pydantic Model) for metadata, and it would be accessible from the asset context (e.g. `context.partition_metadata` or `context.run_config`). When creating RunRequests (i.e., from a sensor), a partition would still be defined with a partition_key, and could also take an optional metadata parameter (or maybe `run_config` could be used?).

**Comment by sryza** [comment]
@spenczar also had a similar request

**Reply by f-tremblay** [reply]
hey sandy, I'm not able to find it (I searched in Discussions, Issues, and Slack). Could you share the link?

**Reply by sryza** [reply]
Ah - I don't think there's documentation for his request.  I just recall it from a direct discussion I had with him.

**Comment by chrishiste** [comment]
@sryza Do you know if this is possible now? 
I'm encoding a lot of metadata in a partition key but the length is so long that it doesn't work anymore.

**Reply by sryza** [reply]
Hey @chrishiste - definitely a reasonable request, but this is still not currently possible

**Reply by chrishiste** [reply]
Alright. Thanks for the answer. I will use a data store instead.

**Comment by stephendeweyEQ** [comment]
Yeah I'd like this as well. The problem I run into is that if you have a Python function in a sensor or schedule which produces dynamic partitions and kicks them off, you may want to attach additional information to the partition as the OP says. You could pass data to the RunRequest that the sensor yields, but if you do that I think it doesn't stick around if you need to rerun the partition without the sensor. As a result I have a pretty wonky setup where I re-derive the metadata in my first asset, even though I shouldn't really need to (because it was already acquired when the partition was created).
